core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2032,4 +2001,207 @@ class ReplicaManager(val config: KafkaConfig,
+          case e: KafkaStorageException =>
+            stateChangeLogger.info(s"Skipped the become-leader state change for ${tp} " +
+              s"with topic id ${info.topicId} due to disk error ${e}")
+            val dirOpt = getLogDir(tp)
+            error(s"Error while making broker the leader for partition ${tp} in dir " +
+              s"${dirOpt}", e)
+        }
+      }
+    }
+  }
+
+  private def applyLocalFollowersDelta(changedPartitions: mutable.HashSet[Partition],
+                                       newImage: MetadataImage,
+                                       delta: TopicsDelta,
+                                       offsetCheckpoints: OffsetCheckpoints,
+                                       newLocalFollowers: mutable.HashMap[TopicPartition, LocalLeaderInfo]): Unit = {
+    stateChangeLogger.info(s"Transitioning ${newLocalFollowers.size} partition(s) to " +
+      "local followers.")
+    replicaFetcherManager.removeFetcherForPartitions(newLocalFollowers.keySet)
+    val shuttingDown = isShuttingDown.get()
+    val partitionsToMakeFollower = new mutable.HashMap[TopicPartition, InitialFetchState]
+    val newFollowerTopicSet = new mutable.HashSet[String]
+    newLocalFollowers.forKeyValue { case (tp, info) =>
+      getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
+        try {
+          newFollowerTopicSet.add(tp.topic())
+
+          completeDelayedFetchOrProduceRequests(tp)
+
+          // Create the local replica even if the leader is unavailable. This is required
+          // to ensure that we include the partition's high watermark in the checkpoint
+          // file (see KAFKA-1647)
+          partition.createLogIfNotExists(isNew, false, offsetCheckpoints, Some(info.topicId))
+
+          if (shuttingDown) {
+            stateChangeLogger.trace(s"Unable to start fetching ${tp} with topic " +
+              s"ID ${info.topicId} because the replica manager is shutting down.")
+          } else {
+            val listenerName = config.interBrokerListenerName.value()
+            val leader = info.partition.leader
+            Option(newImage.cluster().broker(leader)).flatMap(_.node(listenerName).asScala) match {
+              case None => stateChangeLogger.trace(s"Unable to start fetching ${tp} " +
+                s"with topic ID ${info.topicId} from leader ${leader} because it is not " +
+                "alive.")
+              case Some(node) =>
+                val leaderEndPoint = new BrokerEndPoint(node.id(), node.host(), node.port())
+                val log = partition.localLogOrException
+                val fetchOffset = initialFetchOffset(log)
+                partitionsToMakeFollower.put(tp,
+                  InitialFetchState(leaderEndPoint, partition.getLeaderEpoch, fetchOffset))
+            }
+          }
+          changedPartitions.add(partition)
+        } catch {
+          case e: Throwable => stateChangeLogger.error(s"Unable to start fetching ${tp} " +
```

Comment:
Perhaps we should consider adding this partition to the failed partitions that are tracked in the ReplicaFetcherManager.

Reply:
Seems reasonable. It does say the partitions that had "unexpected errors" should end up there.

I'll add it.